c2rust-analyze/src/rewrite/apply.rs

```diff
@@ -0,0 +1,506 @@
+
+            self.emit_bytes(pos, rt.span.lo());
+            self.emit_rewrite(
+                &rt.rw,
+                0,
+                &mut |slf| slf.emit_span_with_rewrites(rt.span, &rt.children),
+                &mut |slf, subexpr_span| slf.emit_span_with_rewrites(subexpr_span, &rt.children),
+            );
+            pos = rt.span.hi();
+        }
+
+        self.emit_bytes(pos, span.hi());
+    }
+}
+
+/// Apply rewrites `rws` to the source files covered by their `Span`s.  Returns a map giving the
+/// rewritten source code for each file that contains at least one rewritten `Span`.
+pub fn apply_rewrites(
+    source_map: &SourceMap,
+    rws: Vec<(Span, Rewrite)>,
+) -> HashMap<FileName, String> {
+    let (rts, errs) = RewriteTree::build(rws);
+    for (span, rw, err) in errs {
+        eprintln!(
+            "{:?}: warning: failed to apply rewrite {:?}: {:?}",
+            span, rw, err
+        );
+    }
+
+    let mut new_src = HashMap::new();
+    let mut rts = &rts as &[RewriteTree<Span>];
+    while rts.len() > 0 {
+        let file = source_map.lookup_source_file(rts[0].span.lo());
+        let idx = rts
+            .iter()
+            .position(|rt| rt.span.lo() >= file.end_pos)
+            .unwrap_or(rts.len());
+        assert!(idx > 0);
+        let (file_rts, rest) = rts.split_at(idx);
+        rts = rest;
+
+        let mut buf = String::new();
+        let mut emit = |s: &str| {
+            buf.push_str(s);
+        };
+        let mut emitter = Emitter {
+            file: &file,
+            emit: &mut emit,
+        };
+        let file_span = Span::new(file.start_pos, file.end_pos, SyntaxContext::root(), None);
+        emitter.emit_span_with_rewrites(file_span, file_rts);
+
+        new_src.insert(file.name.clone(), buf);
+    }
+
+    new_src
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    type FakeSpan = (u32, u32);
+
+    fn mk(start: u32, end: u32, i: usize) -> (FakeSpan, Rewrite<FakeSpan>) {
+        let span = (start, end);
+        let rw = mk_rewrite(i);
+        (span, rw)
+    }
+
+    fn mk_rewrite(i: usize) -> Rewrite<FakeSpan> {
+        Rewrite::Sub(i, (0, 0))
+    }
+
+    fn mk_rt(
+        start: u32,
+        end: u32,
+        i: usize,
+        children: Vec<RewriteTree<FakeSpan>>,
+    ) -> RewriteTree<FakeSpan> {
+        let span = (start, end);
+        let rw = mk_rewrite(i);
+        RewriteTree { span, rw, children }
+    }
+
+    #[test]
+    fn rewrite_tree_nesting() {
+        let (rts, errs) = RewriteTree::build(vec![mk(1, 2, 0), mk(3, 4, 1), mk(0, 5, 2)]);
+        assert_eq!(errs, vec![]);
+        assert_eq!(
+            rts,
+            vec![mk_rt(
+                0,
+                5,
+                2,
+                vec![mk_rt(1, 2, 0, vec![]), mk_rt(3, 4, 1, vec![]),]
+            ),]
+        );
+    }
+
+    #[test]
+    fn rewrite_tree_identical() {
+        let (rts, errs) = RewriteTree::build(vec![
```

Comment:
sorting + nub? I think a comment would help

Reply:
Added a comment explaining the test.